src/main/java/net/neoforged/neoforge/common/crafting/SizedIngredient.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.crafting;
+
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import net.minecraft.network.RegistryFriendlyByteBuf;
+import net.minecraft.network.codec.ByteBufCodecs;
+import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.util.ExtraCodecs;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.Ingredient;
+import net.neoforged.neoforge.common.util.NeoForgeExtraCodecs;
+
+/**
+ * Standard implementation for an ingredient and a count.
+ * {@link Ingredient} does not perform count checks, so this class is used to wrap an ingredient with a count,
+ * and provide a standard serialization format.
+ */
+public record SizedIngredient(Ingredient ingredient, int count) {
```

Comment:
To fully replace my sized ingredient, we would need a method to get the display stacks. Just map the ingredient stacks and set the count. Perhaps lazily cache it, a trick I've used in the past is invalidating the cache if the ingredient list is not reference equal to the last list we saw, though we could also just assume no one is statically storing sized ingredients and thus no need for cache invalidation.

Reply:
Caching is a good idea. We don't have invalidation anymore. (But might add it back since the underlying is not fixed yet...)